src/cmap/establish/handshake.rs

```diff
@@ -38,6 +38,48 @@ pub(crate) struct ClientMetadata {
     pub(crate) os: OsMetadata,
     pub(crate) platform: String,
     pub(crate) env: Option<RuntimeEnvironment>,
+    pub(crate) appended: HashSet<DriverInfo>,
+}
+
+impl ClientMetadata {
+    pub(crate) fn append(&mut self, driver_info: DriverInfo) {
```

Comment:
The spec has the following requirement:

> All strings provided as part of the driver info MUST NOT contain the delimiter used for metadata concatention. Drivers MUST throw an error if any of these strings contains that character.

So we should be checking for "|" in these strings - I think we're also missing that for metadata included in `ClientOptions`.